core/types.go

```diff
@@ -436,11 +436,39 @@ func (s Slot) Next() Slot {
 	}
 }
 
+// Previous returns the previous slot.
+func (s Slot) Previous() Slot {
+	return Slot{
+		Slot:          s.Slot - 1,
+		Time:          s.Time.Add(-1 * s.SlotDuration),
+		SlotsPerEpoch: s.SlotsPerEpoch,
+		SlotDuration:  s.SlotDuration,
+	}
+}
+
 // Epoch returns the epoch of the slot.
 func (s Slot) Epoch() int64 {
 	return s.Slot / s.SlotsPerEpoch
 }
 
+// NextEpoch returns the next epoch of the slot.
+func (s Slot) NextEpoch() int64 {
+	return s.Epoch() + 1
+}
+
+// LastSlotInEpoch returns the last slot in the epoch.
+func (s Slot) LastSlotInEpoch() Slot {
+	lastSlot := s.SlotsPerEpoch*s.NextEpoch() - 1
```

Comment:
this is a bug, you subtract 1 from an epoch, then assign it to a slot variable. Please remember to add tests for logic you add to avoid bugs.

Reply:
Rather remove this. The existing API supports it already:
```
epoch := slot.Epoch()
for s := slot; s.Epoch() == epoch; s = s.Next() {
  ...
}
```

Reply:
> this is a bug, you subtract 1 from an epoch, then assign it to a slot variable. Please remember to add tests for logic you add to avoid bugs.

nope, i don't think it is a bug. we don't subtract 1 from an epoch, we do: `first slot in the next epoch - 1`. Although, it seems we can already iterate using the method you suggested